src/game.rs

```diff
@@ -21,90 +21,248 @@ pub enum ChessError {
     EmptyPieceMove,
 }
 
-#[derive(Clone, Default, PartialEq)]
+#[derive(Clone, Copy, Default, PartialEq, Debug)]
 pub enum GameStatus {
     #[default]
     Ongoing,
     Stalemate,
     Check,
     Checkmate,
+    Replay,
+}
+
+impl GameStatus {
+    pub fn update(&mut self, status: GameStatus) {
+        if *self != status {
+            println!("GameStatus changing from {:?} to {:?}", *self, status)
+        }
+        *self = status;
+    }
 }
 
 #[derive(Clone, Default)]
+pub struct History {
+    history: Vec<(BoardState, Move)>,
+    current_turn: usize,
+}
+
+impl History {
+    pub fn with_state(state: BoardState) -> Self {
+        Self {
+            history: vec![(state, Move::default())],
+            ..Default::default()
+        }
+    }
+
+    fn add_info(&mut self, next_state: BoardState, mv: Move) {
+        self.history.push((next_state, mv));
+        self.current_turn += 1
+    }
+
+    fn get_current_state(&self) -> &BoardState {
+        &self.history[self.current_turn - 1].0
+    }
+
+    fn clone_current_state(&self) -> BoardState {
+        self.get_current_state().clone()
+    }
+
+    fn get_info_for_turn(&self, turn: usize) -> &(BoardState, Move) {
+        &self.history[turn]
+    }
+
+    fn resume(&mut self) {
+        self.current_turn = self.history.len()
+    }
+
+    fn previous_state(&mut self) {
+        if self.current_turn > 1 {
+            self.current_turn -= 1
+        }
+    }
+
+    fn next_state(&mut self) {
+        if self.current_turn < self.history.len() {
+            self.current_turn += 1
+        }
+    }
+
+    fn initial_state(&mut self) {
+        self.current_turn = 1
+    }
+
+    fn is_ongoing(&mut self) -> bool {
+        self.current_turn == self.history.len()
+    }
+}
+
+#[derive(Clone)]
 pub struct Game {
-    state: BoardState,
     valid_moves: HashSet<Move>,
     status: GameStatus,
+    history: History,
+}
+
+impl Default for Game {
+    fn default() -> Self {
+        Self::with_state(BoardState::default())
+    }
 }
 
 impl Game {
     pub fn new() -> Self {
-        let mut game = Self::default();
+        Self::default()
+    }
+
+    pub fn with_history(history: History) -> Self {
+        let mut game = Self {
+            history,
+            ..Default::default()
```

Comment:
I think this causes an issue when it ultimately gets called inside `Game::default()`